server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java

```diff
@@ -0,0 +1,145 @@
+import com.generalbytes.batm.common.currencies.FiatCurrency;
+import com.generalbytes.batm.server.extensions.*;
+import com.generalbytes.batm.server.extensions.extra.bitcoin.exchanges.binance.BinanceComExchange;
+import com.generalbytes.batm.server.extensions.extra.bitcoin.exchanges.binance.BinanceUsExchange;
+import com.generalbytes.batm.server.extensions.extra.bitcoin.sources.coingecko.CoinGeckoRateSource;
+import com.generalbytes.batm.server.extensions.extra.bitcoin.sources.coinpaprika.CoinPaprikaRateSource;
+import com.generalbytes.batm.server.extensions.extra.dash.sources.coinmarketcap.CoinmarketcapRateSource;
+import com.generalbytes.batm.server.extensions.extra.nano.wallet.demo.DemoWallet;
+import com.generalbytes.batm.server.extensions.extra.nano.wallet.node.NanoNodeWallet;
+import com.generalbytes.batm.server.extensions.extra.nano.wallet.paper.NanoPaperWalletGenerator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.StringTokenizer;
+
+/*
+ * EXTENSION NOTES:
+ *
+ * The NanoExtensionContext object which is passed around contains various different objects relating to the
+ * cryptocurrency being used, including the crypto code identifier, address format and standard unit denomination.
+ *
+ * The context object also contains a reference to a globally-used NanoRpcClient instance. This is optionally
+ * defined when the end user configures a nano_node wallet — if this is not defined, then certain actions will
+ * become unavailable (or resort to less desirable fallbacks). This is a (messy) necessity due to the dependency
+ * constraints put in place, so operations such as paper wallet generation and address validation must be processed
+ * externally on an RPC endpoint.
+ */
+public class NanoExtension extends AbstractExtension {
+
+    private static final Logger log = LoggerFactory.getLogger(NanoExtension.class);
+
+    public static final CryptoCurrency CRYPTO = CryptoCurrency.NANO;
+
+    private volatile NanoExtensionContext context = new NanoExtensionContext(CRYPTO, ctx, NanoCurrencyUtil.NANO);
+
+
+    @Override
+    public String getName() {
+        return "BATM Nano extra extension";
+    }
+
+    @Override
+    public void init(IExtensionContext ctx) {
+        super.init(ctx);
+        this.context = new NanoExtensionContext(CRYPTO, ctx, NanoCurrencyUtil.NANO);
+    }
+
+    @Override
+    public IWallet createWallet(String walletLogin, String tunnelPassword) {
+        try {
+            if (walletLogin != null && !walletLogin.trim().isEmpty()) {
+                StringTokenizer st = new StringTokenizer(walletLogin, ":");
+                String walletName = st.nextToken();
+
+                if ("nano_node".equalsIgnoreCase(walletName)) {
+                    NanoNodeWallet wallet = NanoNodeWallet.create(context, st);
+                    context.setRpcClient(wallet.getRpcClient()); // Set global RPC client
```

Comment:
would this be OK if this extension is used by multiple organizations (operators) with multiple different rpc wallets on the same server? Why is this needed?

Reply:
Hey mate, thanks for reviewing the PR!

A 'global' `NanoRpcClient` instance is used as Nano relies on some additional cryptographic functions which were rejected in a former PR as multiple libraries are required — to get around this, I changed it so the node processes all the cryptographic work instead of being done locally on the CAS server. This does however mean that to use basic features like the paper wallet generator, a `nano_node` wallet is still required and needs to be configured by the operator. Not a very pretty way of doing things I'll admit, but it's the only solution I could think up at the time.

After reading some of the documentation, I was under the impression that each organisation would use their own independent CAS server (and Nano node) rather than being a shared system — if this isn't the case or there are other security considerations that need to be make, please do let me know and I'll review what the options are.

Reply:
The operators can choose to host their own CAS server (and in that case they probably would have just one Nano node) but if we merge this PR into our release it will also be available for all the operators using our "cloud" hosted CAS server that is shared between different operators. In that case everything has to be completely separate for each of them. 

Would a more basic implementation without the global RPC node work? For example without the support for paper wallets or just a basic address validation? Or is paper wallet support etc really required? What other features depend on it?

Another option would be to keep the extension private (not merged in our product) and you (the operator?) would install it only on their own server.

Reply:
Ah I see, I didn't realise that was an option. With the cloud CAS service, would the Nano node be ran by yourselves and shared, or would each ATM operator configure and run their own node?

As for support of paper wallets, I'm honestly not sure if it's a requirement. I was under the impression that paper wallets are an essential feature to purchase the crypto, but I honestly haven't seen one of your ATMs in use so it's a bit hard to judge what features may be required.

Could you explain further on how to install the extension privately? That could work for the time being, as we have someone wanting to get this up and running ASAP. I always figured the CAS system would require signing of the JARs as a security protocol, so never thought of it as an option.